File: python/MinStack.py
class MinStack(object):

    def __init__(self):
       self.stack = list()

    def push(self, number):
        self.stack.append((number, number if 0 == len(self.stack) else min(number, self.stack[-1][1])))

    def pop(self):
        return self.stack.pop()[0]

    def min(self):
        return self.stack[-1][1]

File: scala/MinStack.scala
package leetcode

import scala.math.min

//https://leetcode.com/problems/min-stack/
class MinStack {
    /** initialize your data structure here. */
    
    private var stk: List[(Int, Int)] = List()
  
    def push(x: Int) {
        stk = (if(true == stk.isEmpty) x else min(x, stk(0)._1), x)::stk
    }
  
    def pop() {
        stk = stk.drop(1)
    }
  
    def top(): Int = {
        stk(0)._2 
    }
    
    def getMin(): Int = {
        stk(0)._1
    }
}
